### PR TITLE
ContactField dependencies

### PR DESCRIPTION
**Pull request by ericnewcomer**
* Adds contact fields (as type string) when importing flows if they don't exist, even if referenced as part of an expression
* Creates dependencies on contact fields when saving a flow
* Parses expressions in the editor and disallows if they reference contact fields that don't exist

### CODE REVIEW

File: temba/contacts/models.py [hunk 1]
```diff
@@ -365,6 +365,10 @@ def is_valid_label(cls, label):
     def hide_field(cls, org, user, key):
         existing = ContactField.objects.filter(org=org, key=key).first()
         if existing:
+            from temba.flows.models import Flow
+            if Flow.objects.filter(field_dependencies__in=[existing]).first():
+                raise Exception(_("Cannot delete field '%s' while used in flows." % key))
```
Comment (rowanseymour) [inline temba/contacts/models.py]: Where we use exceptions as runtime assertions we tend to use `ValueError` and there's not much point in using gettext

Comment (rowanseymour) [reply temba/contacts/models.py]: Also better to use `.exists()` if we're just checking existence

File: temba/flows/migrations/0114_flow_field_dependencies.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    dependencies = [
+        ('contacts', '0067_auto_20170808_1852'),
+        ('flows', '0113_auto_20171012_1839'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='flow',
+            name='field_dependencies',
+            field=models.ManyToManyField(blank=True, help_text='Any fields this flow depends on', related_name='dependent_fields', to='contacts.ContactField', verbose_name=''),
```
Comment (rowanseymour) [inline temba/flows/migrations/0114_flow_field_dependencies.py]: think related_name should be `dependent_flows`

File: temba/utils/expressions.py [hunk 3]
```diff
@@ -257,3 +259,35 @@ def replace_embedded_filter_style(match):
         expression = '(%s)' % expression
 
     return expression
+
+
+class ContactFieldCollector (EvaluationContext):
```
Comment (rowanseymour) [inline temba/utils/expressions.py]: PEP didn't complain about this space ??

File: templates/partials/split_editor.haml [hunk 4]
```diff
@@ -104,13 +114,14 @@
                select-server:"/api/resthooks/?_format=select2", type:"hidden"}
 
     .webhook{ng-if:'formData.rulesetConfig.type == "webhook"'}><
-      .split-margin
+      .split-margin
```
Comment (rowanseymour) [inline templates/partials/split_editor.haml]: vscode got a setting to trim end of line spaces?

**Comment by nicpottier** [comment]
Would be good to merge this as I could really use the dependencies for wrapping contact date values.

**Pull request merged**